### ISSUE TITLE
Axis autoscaling doesn't take text into account

### ISSUE THREAD

**Issue by Phlya** [matplotlib/matplotlib#10497]
### Bug report

**Bug summary**

Autoscaling of axes doesn't take location of `text`s and `annotation`s into account - so a plot with only text defaults to limits (0, 1) and (0, 1), and if the texts are actually outside these coordinates, the plot will be blank with only axes.

**Code for reproduction**

```python
plt.subplots()
plt.annotate(xy=(3, 3), s='test')
```

**Actual outcome**

![image](https://user-images.githubusercontent.com/2895034/36320549-81372726-133e-11e8-9e8f-7df06cf639ed.png)

**Expected outcome**
Something like this, perhaps?
Adding, for example:
```
plt.xlim(2, 4)
plt.ylim(2, 4)
```
![image](https://user-images.githubusercontent.com/2895034/36320598-9adbd28a-133e-11e8-9b78-20b55cc04aee.png)

**Matplotlib version**

  * Operating system: Ubuntu 16.04
  * Matplotlib version: 2.1.0
  * Matplotlib backend (`print(matplotlib.get_backend())`): nbAgg
  * Python version: 3.5.2

Want to mention, this complicates using `adjustText` (https://github.com/Phlya/adjustText) since it has to be called only after some other plotting has been done and the axes are set up to correct limits.

**Comment by anntzer** [comment]
You can (should) call `fig.canvas.draw()` to get the correct limits before running your adjustText algorithm.

Text rendering is only one of many relatively costly operations that are deferred until draw time (and therefore, as in this case, not taken into account during autoscaling).  Actually per #7413 autoscaling itself should probably be deferred to draw time...

Suggesting to close as wontfix.

**Comment by Phlya** [comment]
So, if autoscaling is changed to work at draw time, will this get easily fixed?

And adjustText works without calling `draw()`, except when it's called on `basemap` axes, but it does use the renderer from the figure.

**Comment by anntzer** [comment]
Moving autoscaling to draw time is likely to require some major surgery, but is also likely (well, can't say for sure) to help this issue (and others).

**Comment by vak** [comment]
is it only text-related  issue or the same for other entities?

```
fig, ax = plt.subplots()
patches = []
patches.append(Polygon(np.array([[0,0], [0,0.6], [1.3,2.0]]), True))
colors = 100*np.random.rand(len(patches))
p = PatchCollection(patches, alpha=0.4)
p.set_array(np.array(colors))
ax.add_collection(p)
fig.colorbar(p, ax=ax)
plt.show()
```
![image](https://user-images.githubusercontent.com/69862/36641221-56e221ac-1a2c-11e8-9281-d6f804c4cebf.png)

**Comment by 2xB** [comment]
Interestingly enough, **in jupyter notebook** (v5.2.2, running on kde neon, Python 3.6.7, Matplotlib v3.0.3) the figure is drawn as shown by @Phlya, but the canvas itself is resized correctly, causing the figure to only take a small part of the canvas. For example, running the first example
```
import matplotlib.pyplot as plt
%matplotlib inline
plt.annotate(xy=(3, 3), s='test')
```
leads to:

---
![grafik](https://user-images.githubusercontent.com/31772910/58486960-2098a500-8167-11e9-9ed2-3485d84e7c0d.png)
---

**Comment by jklymak** [comment]
Inline saves the figure w bbox_inches=tight which includes all items added to the figure. 

I don’t see any reason autolim couldn’t be called on the position of the text, except that would be a pretty big change of behaviour.  However I would argue quite strenuously that it should never be called to make the limits fit the full rendered extent of the text.  Setting data limits based on the physical size of a draw element requires an inverse problem to be solved and it’s not really worth the complication.

**Comment by 2xB** [comment]
@jklymak Disclamer: I am not familiar with the matplotlib code.

If ˋbbox_inches=tightˋ includes all items added to the figure, then why does a scatterplot with data points outside the plotting region (set with plt.xlim) does not lead to this behaviour of large whitespace on the canvas outside the plot itself?

**Comment by jklymak** [comment]
Because markers are clipped. If you clip the annotation it should not be included either... If it is, we should open a bug...
